File: python/evm_state/reader.py
"""Pin an immutable checkpoint and page its complete nonzero storage."""
import base64
import json
import os
import time
import uuid

from .checkpoint import _manifest, _read_account
from .control import control, object_id
from .files import atomic_json
from .proof import VerificationError, address, unhex


def pin(client, snapshot_id, purpose="reader"):
    owner = control(client)
    with owner.reader():
        ready = _manifest(client, snapshot_id)
        result = {"pin_id": uuid.uuid4().hex, "snapshot_id": snapshot_id, "control_id": owner.record["control_id"],
                  "created_at": time.time_ns(), "purpose": purpose, "header": ready["header"]}
        atomic_json(owner.pins / (result["pin_id"] + ".json"), result)
        return result


def _pin(owner, pin_id):
    record = json.loads((owner.pins / (object_id(pin_id) + ".json")).read_text())
    if record.get("pin_id") != pin_id or record.get("control_id") != owner.record["control_id"]:
        raise VerificationError("pin belongs to another controller or is corrupt")
    object_id(record["snapshot_id"])
    return record


def unpin(client, pin_id):
    owner = control(client)
    with owner.reader():
        record = _pin(owner, pin_id)
        (owner.pins / (pin_id + ".json")).unlink()
        fd = os.open(owner.pins, os.O_RDONLY)
        try:
            os.fsync(fd)
        finally:
            os.close(fd)
        return {"pin_id": pin_id, "snapshot_id": record["snapshot_id"], "released": True}


def list_pins(client):
    owner = control(client)
    with owner.reader():
        return sorted((_pin(owner, path.stem) for path in owner.pins.glob("*.json")), key=lambda value: value["created_at"])


def _encode_cursor(snapshot_id, account, slot):
    value = {"version": 1, "snapshot_id": snapshot_id, "address": account, "after_slot": slot}
    return base64.urlsafe_b64encode(json.dumps(value, sort_keys=True, separators=(",", ":")).encode()).decode()


def _decode_cursor(token, snapshot_id, account):
    try:
        if len(token) > 1024:
            raise ValueError("oversized cursor")
        value = json.loads(base64.b64decode(token, altchars=b"-_", validate=True))
        if value.get("version") != 1 or value.get("snapshot_id") != snapshot_id or value.get("address") != account:
            raise ValueError("cursor belongs to another checkpoint or account")
        unhex(value["after_slot"], 32)
        return value["after_slot"]
    except (ValueError, KeyError, TypeError) as error:
        raise VerificationError("invalid or mismatched storage cursor") from error


def page(client, pin_id, account, cursor=None, limit=1000):
    if isinstance(limit, bool) or not 1 <= limit <= 10000:
        raise ValueError("page limit must be between 1 and 10000")
    owner = control(client)
    account = address(account)
    with owner.reader():
        pinned = _pin(owner, pin_id)
        snapshot_id = pinned["snapshot_id"]
        metadata = _read_account(client, snapshot_id, account)
        metadata["nonce"] = str(int(metadata["nonce"]))
        after = _decode_cursor(cursor, snapshot_id, account) if cursor else ""
        rows = list(client.rows("SELECT slot,value FROM checkpoint_storage FINAL WHERE snapshot_id={id:String} "
            "AND address={address:String} AND slot>{after:String} ORDER BY slot LIMIT {limit:UInt32}",
            {"id": snapshot_id, "address": account, "after": after, "limit": limit + 1}))
        has_more = len(rows) > limit
        rows = rows[:limit]
        return {"pin_id": pin_id, "snapshot_id": snapshot_id, "header": metadata["header"], "account": metadata,
                "storage": rows, "next_cursor": _encode_cursor(snapshot_id, account, rows[-1]["slot"]) if has_more else None}

File: crates/evm-state/src/reader.rs
//! Durable pins and account-bound keyset cursors over immutable checkpoints.
use crate::{
    ch::{params, uint, ClickHouse},
    checkpoint::{manifest_unlocked, read_account_unlocked},
    control::{new_id, object_id, Control},
    files::{atomic_json, canonical_json},
    proof::{address, fixed, string},
};
use anyhow::{ensure, Context, Result};
use base64::{engine::general_purpose::URL_SAFE, Engine};
use serde_json::{json, Value};
use std::{
    fs,
    time::{SystemTime, UNIX_EPOCH},
};

pub fn now_ns() -> Result<u64> {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)?
        .as_nanos()
        .try_into()
        .context("timestamp exceeds uint64")
}

pub fn pin(client: &ClickHouse, snapshot_id: &str, purpose: &str) -> Result<Value> {
    let owner = Control::open(client)?;
    let _reader = owner.reader()?;
    let ready = manifest_unlocked(client, snapshot_id)?;
    let result = json!({"pin_id":new_id(),"snapshot_id":snapshot_id,"control_id":owner.record["control_id"],
        "created_at":now_ns()?,"purpose":purpose,"header":ready["header"]});
    atomic_json(
        &owner
            .pins
            .join(format!("{}.json", string(&result, "pin_id")?)),
        &result,
        false,
    )?;
    Ok(result)
}

pub(crate) fn read_pin(owner: &Control, pin_id: &str) -> Result<Value> {
    let record: Value = serde_json::from_slice(&fs::read(
        owner.pins.join(format!("{}.json", object_id(pin_id)?)),
    )?)?;
    ensure!(
        record["pin_id"] == pin_id && record["control_id"] == owner.record["control_id"],
        "pin belongs to another controller or is corrupt"
    );
    object_id(string(&record, "snapshot_id")?)?;
    Ok(record)
}

pub fn unpin(client: &ClickHouse, pin_id: &str) -> Result<Value> {
    let owner = Control::open(client)?;
    let _reader = owner.reader()?;
    let record = read_pin(&owner, pin_id)?;
    fs::remove_file(owner.pins.join(format!("{pin_id}.json")))?;
    fs::File::open(&owner.pins)?.sync_all()?;
    Ok(json!({"pin_id":pin_id,"snapshot_id":record["snapshot_id"],"released":true}))
}

pub fn list_pins(client: &ClickHouse) -> Result<Vec<Value>> {
    let owner = Control::open(client)?;
    let _reader = owner.reader()?;
    pins_unlocked(&owner)
}

pub(crate) fn pins_unlocked(owner: &Control) -> Result<Vec<Value>> {
    let mut pins = Vec::new();
    for entry in fs::read_dir(&owner.pins)? {
        let path = entry?.path();
        if path.extension().is_some_and(|v| v == "json") {
            let record = read_pin(
                owner,
                path.file_stem()
                    .and_then(|v| v.to_str())
                    .context("invalid pin file")?,
            )?;
            let created = uint(&record["created_at"])?;
            pins.push((created, record));
        }
    }
    pins.sort_by_key(|(time, _)| *time);
    Ok(pins.into_iter().map(|(_, record)| record).collect())
}

pub fn encode_cursor(snapshot_id: &str, account: &str, slot: &str) -> Result<String> {
    Ok(URL_SAFE.encode(canonical_json(
        &json!({"version":1,"snapshot_id":snapshot_id,"address":account,"after_slot":slot}),
    )?))
}

pub fn decode_cursor(token: &str, snapshot_id: &str, account: &str) -> Result<String> {
    ensure!(token.len() <= 1024, "oversized storage cursor");
    let value: Value =
        serde_json::from_slice(&URL_SAFE.decode(token).context("invalid storage cursor")?)?;
    ensure!(
        value["version"] == 1 && value["snapshot_id"] == snapshot_id && value["address"] == account,
        "cursor belongs to another checkpoint or account"
    );
    let slot = string(&value, "after_slot")?;
    fixed::<32>(slot)?;
    Ok(slot.into())
}

pub fn page(
    client: &ClickHouse,
    pin_id: &str,
    selected: &str,
    cursor: Option<&str>,
    limit: usize,
) -> Result<Value> {
    ensure!(
        (1..=10000).contains(&limit),
        "page limit must be between 1 and 10000"
    );
    let owner = Control::open(client)?;
    let selected = address(selected)?;
    let _reader = owner.reader()?;
    let pinned = read_pin(&owner, pin_id)?;
    let snapshot_id = string(&pinned, "snapshot_id")?;
    let mut metadata = read_account_unlocked(client, snapshot_id, &selected)?;
    metadata["nonce"] = json!(uint(&metadata["nonce"])?.to_string());
    let after = cursor
        .filter(|s| !s.is_empty())
        .map(|s| decode_cursor(s, snapshot_id, &selected))
        .transpose()?
        .unwrap_or_default();
    let mut rows = client.rows("SELECT slot,value FROM checkpoint_storage FINAL WHERE snapshot_id={id:String} AND address={address:String} AND slot>{after:String} ORDER BY slot LIMIT {limit:UInt32}",
        &params(json!({"id":snapshot_id,"address":selected,"after":after,"limit":limit+1}))?)?.collect::<Result<Vec<_>>>()?;
    let more = rows.len() > limit;
    rows.truncate(limit);
    let next = if more {
        Some(encode_cursor(
            snapshot_id,
            &selected,
            string(rows.last().context("empty nonterminal page")?, "slot")?,
        )?)
    } else {
        None
    };
    Ok(
        json!({"pin_id":pin_id,"snapshot_id":snapshot_id,"header":metadata["header"],"account":metadata,"storage":rows,"next_cursor":next}),
    )
}
